gitpod-io/gitpod
# Make TakeSnapshot more reliable

## Description
This fixes #5862 by following this idea: https://github.com/gitpod-io/gitpod/issues/5862#issuecomment-946721722

### In detail
 - we add a `state: 'pending' | 'error' | 'available''` field to `Snapshot` (+ `message` and `availableTime`)
 - we add a method `content-service.WorkspaceSnapshotExists`
 - we make sure `ws-daemon.TakeSnapshot` returns early
 - we add an API method `server.waitForSnapshot`
 - we add a `SnapshotService` to `server` which is responsible for making sure all 
 - we change the semantics fo `server.takeSnapshot` to:
   - return as early as we have the upload file URL from `ws-daemon`, stored that in the DB (with state: 'pending') and return the `snapshotId`
   - we make sure `SnapshotService` monitors this DB entry and regularly checks if it's done or timed out, resulting in an update to the DB
   - clients have to take that `snapshotId` and have to call `server.waitForSnapshot` to wait for the snapshot to become `available` (or `error` in case sth went south). This internally waits for the same Promise `SnapshotService` is already driving

### Regarding deployment
This change has breaking changes to IDE and workspace. To untangle and make the deployment as easy as possible:
 - once reviewed (and tested) split this PR into two: IDE and workspace + meta
 - `server` and workspace are fully backwards-compatible to each other; `server` API is backwards compatible as well. Thus we have to make sure meta is deployed before IDE, the rest does not matter.
 
 

## Related Issue(s)

Fixes #5862

IDE companion PR: https://github.com/gitpod-io/openvscode-server/pull/169

## How to test
- open a dev workspace on this branch: https://gitpod.io/#https://github.com/gitpod-io/gitpod/pull/6144
- **prepare** to execute `kubectl rollout restart deployment proxy`
-  start a test workspace with this link: https://gpl-5862-snapshot.staging.gitpod-dev.com/#https://github.com/shaal/gitpod-test-failing-snapshot (thx @shaal ! :pray: )
- F1 -> snapshot -> enter
- **execute** `kubectl rollout restart deployment proxy` from the dev workspace (this simulates ws connection loss)
- note how the call fails (shows that it's backwards compatible!)
- ~note how the snapshot process is not interrupted in the IDE of the test workspace~
- ~note that after ~2mins the snapshot is successful~
- ~test the snapshot link and note that it works~

## Release Notes

```release-note
make "snapshots" more reliable
```

## Documentation
